posts/2022-10-28-gats-stabilization.md

```diff
@@ -0,0 +1,275 @@
+
+Here, we'll go over just a couple of the limitations that we've identified that users might run into.
+
+### Implied `'static` requirement from higher-ranked trait bounds
+
+Consider the following code:
+
+```rust
+trait LendingIterator {
+    type Item<'a> where Self: 'a;
+}
+
+pub struct WindowsMut<'x, T> {
+    slice: &'x [T],
+}
+
+impl<'x, T> LendingIterator for WindowsMut<'x, T> {
+    type Item<'a> = &'a mut [T] where Self: 'a;
+}
+
+fn print_items<I>(iter: I)
+where
+    I: LendingIterator,
+    for<'a> I::Item<'a>: Debug,
+{ ... }
+
+fn main() {
+    let mut array = [0; 16];
+    let slice = &mut array;
+    let windows = WindowsMut { slice };
+    print_items::<WindowsMut<'_, usize>>(windows);
+}
+```
+
+Here, imagine we wanted to have a `LendingIterator` where the items are overlapping slices of an array. We also have a function `print_items` that prints every item of a `LendingIterator`, as long as they implement `Debug`. This all seems innocent enough, but the above code doesn't compile — even though it should. Without going into details here, the `for<'a> I::Item<'a>: Debug` currently implies that `I::Item<'a>` must outlive `'static`.
+
+This is not really a nice bug. And of all the ones we'll mention today, this will likely be the one that is most limiting, annoying, and tough to figure out. This pops up much more often with GATs, but can be found with code that doesn't use GATs at all. Unfortunately, fixing this requires some refactorings to the compiler that isn't a short-term project. It is on the horizon though. The good news is that, in the meantime, we are least working on improving the error message you get from this code. This is what it will look like in the upcoming stabilization:
+
+```rust
+error[E0597]: `array` does not live long enough
+   |
+   |     let slice = &mut array;
+   |                 ^^^^^^^^^^ borrowed value does not live long enough
+   |     let windows = WindowsMut { slice };
+   |     print_items::<WindowsMut<'_, usize>>(windows);
+   |     -------------------------------------------- argument requires that `array` is borrowed for `'static`
+   | }
+   | - `array` dropped here while still borrowed
+   |
+note: due to current limitations in the borrow checker, this implies a `'static` lifetime
+   |
+   |     for<'a> I::Item<'a>: Debug,
+   |                          ^^^^
+```
+
+It's not perfect, but it's something. It might not cover all cases, but if have a `for<'a> I::Item<'a>: Trait` bound somewhere and get an error that says something doesn't live long enough, you might be running into this bug. We're actively working to fix this, but it doesn't seem to come up often enough, so we feel the feature is still immensely useful enough even with it around.
```

Comment:
Wording around "doesn't come up often enough" feels a little off to me. I think it might help to split that sentence up a little and replace the "it" with "this deficiency" or something like that.